cdap/resource_streaming_program.go

```diff
@@ -0,0 +1,248 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cdap
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+)
+
+// https://docs.cdap.io/cdap/current/en/reference-manual/http-restful-api/lifecycle.html.
+func resourceStreamingProgram() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceStreamingProgramCreate,
+		Read:   resourceStreamingProgramRead,
+		Delete: resourceStreamingProgramDelete,
+		Exists: resourceStreamingProgramExists,
+
+		Schema: map[string]*schema.Schema{
+			"namespace": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "The name of the namespace in which this resource belongs. If not provided, the default namespace is used.",
+				DefaultFunc: func() (interface{}, error) {
+					return defaultNamespace, nil
+				},
+			},
+			"app": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Name of the application.",
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Name of the program.",
+				DefaultFunc: func() (interface{}, error) {
+					return "DataStreamsSparkStreaming", nil
+				},
+			},
+			"type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				Description:  "One of flows, mapreduce, services, spark, workers, or workflows.",
+				ValidateFunc: validation.StringInSlice([]string{"flows", "mapreduce", "services", "spark", "workers", "workflows"}, false),
+				DefaultFunc: func() (interface{}, error) {
+					return "spark", nil
+				},
+			},
+			"runtime_arguments": {
+				Type:        schema.TypeMap,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The runtime arguments used to start the program",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+func resourceStreamingProgramCreate(d *schema.ResourceData, m interface{}) error {
+	config := m.(*Config)
+	name := d.Get("name").(string)
+
+	addr := urlJoin(
+		config.host,
+		"/v3/namespaces", d.Get("namespace").(string),
+		"/apps", d.Get("app").(string), d.Get("type").(string),
+		name)
+	startAddr := urlJoin(
+		addr, "start")
+	statusAddr := urlJoin(
+		addr, "status")
+
+	argsObj := make(map[string]interface{})
+
+	b, err := json.Marshal(argsObj)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, startAddr, bytes.NewReader(b))
+	if err != nil {
+		return err
+	}
+
+	if _, err := httpCall(config.httpClient, req); err != nil {
+		return err
+	}
+
+	// Poll until actually reaches RUNNING state.
+	for {
+		p, err := getProgramStatus(config, statusAddr)
+		if err != nil {
+			return err
+		}
+
+		switch s := p.Status; s {
```

Comment:
You can simplify this check to the following:

switch p.Status {
  case "RUNNING":
     return nil
  case "FAILED":
    return resource.NonRetryableError(fmt.Errorf("program in unexpected state: %v", p.Status))
  default:
    return resource.RetryableError(fmt.Errorf("program in state %v", p.Status))
}
I am guessing retry helper also takes care of the sleep?